src/contracts/libraries/GPv2Trade.sol

```diff
@@ -0,0 +1,148 @@
+    ///
+    /// @param trade The trade.
+    /// @param tokens The list of tokens included in the settlement. The token
+    /// indices in the trade parameters map to tokens in this array.
+    /// @param order The memory location to extract the order data to.
+    function extractOrder(
+        Data calldata trade,
+        IERC20[] calldata tokens,
+        GPv2Order.Data memory order
+    ) internal pure returns (GPv2Signing.Scheme signingScheme) {
+        order.sellToken = tokens[trade.sellTokenIndex];
+        order.buyToken = tokens[trade.buyTokenIndex];
+        order.receiver = trade.receiver;
+        order.sellAmount = trade.sellAmount;
+        order.buyAmount = trade.buyAmount;
+        order.validTo = trade.validTo;
+        order.appData = trade.appData;
+        order.feeAmount = trade.feeAmount;
+        (order.kind, order.partiallyFillable, signingScheme) = extractFlags(
+            trade.flags
+        );
+    }
+
+    /// @dev Decodes trade flags.
+    ///
+    /// Trade flags are used to tightly encode information on how to decode
+    /// an order. Examples that directly affect the structure of an order are
+    /// the kind of order (either a sell or a buy order) as well as whether the
+    /// order is partially fillable or if it is a "fill-or-kill" order. It also
+    /// encodes the signature scheme used to validate the order. As the most
+    /// likely values are fill-or-kill sell orders by an externally owned
+    /// account, the flags are chosen such that `0x00` represents this kind of
+    /// order. The flags byte uses the following format:
+    ///
+    /// ```
+    /// bit | 31 ... 4 | 3 | 2 | 1 | 0 |
+    /// ----+-----------------------+---+---+
+    ///     |  unused  | *   * | * | * |
+    ///                  |   |   |   |
+    ///                  |   |   |   +---- order kind bit, 0 for a sell order
+    ///                  |   |   |         and 1 for a buy order
+    ///                  |   |   |
+    ///                  |   |   +-------- order fill bit, 0 for fill-or-kill
+    ///                  |   |             and 1 for a partially fillable order
+    ///                  |   |
+    ///                  +---+------------ signature scheme bits:
+    ///                                    00: EIP-712
+    ///                                    01: eth_sign
+    ///                                    10: EIP-1271
+    ///                                    11: unused
+    /// ```
+    function extractFlags(uint256 flags)
+        internal
+        pure
+        returns (
+            bytes32 kind,
+            bool partiallyFillable,
+            GPv2Signing.Scheme signingScheme
+        )
+    {
+        if (flags & 0x01 == 0) {
+            kind = GPv2Order.SELL;
+        } else {
+            kind = GPv2Order.BUY;
+        }
+        partiallyFillable = flags & 0x02 != 0;
+        signingScheme = GPv2Signing.Scheme(flags >> 2);
```

Comment:
TIL that you can initialize enums like this. I assume this line fails if the value is larger than the number of enums.
The documentation is technically wrong since bits 31 ... 4 are actually signature bits without an `& 3`. :stuck_out_tongue:

Reply:
> The documentation is technically wrong since bits 31 ... 4 are actually signature bits without an & 3

We should say _reserved_ not unused (as it must be 0). I'm just taking advantage of the fact that solidity checks that the value is valid `GPv2Signing.Scheme` instead of checking manually.

Reply:
I added a comment about this as well, so its more obvious.